### PR TITLE
Add code check CI tests, add docs about ReferenceResolution, and other cleanups

### PR DESCRIPTION
**Pull request by natemcmaster**
Changes:
* Add a step which checks that generated code is up to date
* Copy over the documentation on how to work with `<reference>` 
* Fixup a broken reference in the .sln
* Fix the casing on Microsoft.Extensions.ValueStopwatch.Sources
* Remove some unused references and variables

### CODE REVIEW

File: eng/scripts/CodeCheck.ps1 [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        $slnDir = Split-Path -Parent $_
+        $sln = $_
+        & dotnet sln $_ list `
+            | ? { $_ -ne 'Project(s)' -and $_ -ne '----------' } `
+            | % {
+                $proj = Join-Path $slnDir $_
+                if (-not (Test-Path $proj)) {
+                    LogError "Missing project. Solution references a project which does not exist: $proj. [$sln] "
+                }
+            }
```
Comment (dougbu) [inline eng/scripts/CodeCheck.ps1]: Suggest being more consistent with indentation in the two `foreach` blocks. (This is a bit harder to read that in needs to be.) The first `Foreach-Object` block has no indentation but the second has. It's fine if either the close curlies aren't laid out perfectly (as they aren't for the outer line continuation).

FYI I'm going to put `dotnet sln $_ list ...` into a function in my PowerShell profile 😃

File: src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj [hunk 2]
```diff
@@ -3,10 +3,11 @@
   <PropertyGroup>
     <Description>Analyzer package for async best practices.</Description>
     <VersionPrefix>$(ExperimentalVersionPrefix)</VersionPrefix>
-    <VerifyVersion>false</VerifyVersion>
     <TargetFramework>netstandard1.3</TargetFramework>
     <NoWarn>$(NoWarn);CS1591</NoWarn>
     <BuildOutputTargetFolder>analyzers/dotnet/cs/</BuildOutputTargetFolder>
+      <!-- Analyzers cannot be referenced as a ProjectReference. -->
```
Comment (dougbu) [inline src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj]: You've got a `<ProjectReference />` from `test/Microsoft.DotNet.Analyzers.Async.Tests.csproj` to this project. Did you mean `as a Reference`?

File: src/Configuration/Config/test/Microsoft.Extensions.Configuration.Tests.csproj [hunk 3]
```diff
@@ -6,11 +6,11 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\Config.Binder\src\Microsoft.Extensions.Configuration.Binder.csproj" />
+    <ProjectReference Include="$(RepoRoot)src\Configuration\testassets\Test.Common\Microsoft.Extensions.Configuration.Test.Common.csproj" />
```
Comment (dougbu) [inline src/Configuration/Config/test/Microsoft.Extensions.Configuration.Tests.csproj]: Suggest being consistent -- `..\..\testasssets\`

File: src/TestingUtils/Internal.AspNetCore.Analyzers/src/Internal.AspNetCore.Analyzers.csproj [hunk 4]
```diff
@@ -6,6 +6,10 @@
     <PackageTags>$(PackageTags);analyzers</PackageTags>
     <NoWarn>$(NoWarn);CS1591</NoWarn>
     <BuildOutputTargetFolder>analyzers/dotnet/cs/</BuildOutputTargetFolder>
+    <!-- Analyzers cannot be referenced as a ProjectReference. -->
```
Comment (dougbu) [inline src/TestingUtils/Internal.AspNetCore.Analyzers/src/Internal.AspNetCore.Analyzers.csproj]: Again, suggest "as a Reference"

File: eng/scripts/CodeCheck.ps1 (hunk 1 above)
Comment (natemcmaster) [reply eng/scripts/CodeCheck.ps1]: I too which the indentation was better. The auto-formatter in VSCode does this, so I'll manually undo it for now.

File: src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj (hunk 2 above)
Comment (natemcmaster) [reply src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj]: This is deliberate. Analyzers, like MSBuild tasks, are "special package" types. Although they carry assemblies, they don't end up in the `lib/` folder of a NuGet package. As a result, PackageRef and ProjectRef are not interchangeable, so I've made it an explicit ProjectRef.

Comment (natemcmaster) [reply src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj]: I'll update the comment to clarify.

Comment (dougbu) [reply src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj]: I understand the context but saying "PackageReference" here confuses the issue. My thought was just to say "as a Reference".

Comment (natemcmaster) [reply src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj]: I updated the comment. Let me know if the update is clearer.

**Pull request merged**